src/Bicep.Core/Semantics/Namespaces/NamespaceProvider.cs

```diff
@@ -0,0 +1,261 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using Bicep.Core.Configuration;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Extensions;
+using Bicep.Core.Features;
+using Bicep.Core.Syntax;
+using Bicep.Core.Syntax.Visitors;
+using Bicep.Core.TypeSystem;
+using Bicep.Core.TypeSystem.Providers;
+using Bicep.Core.TypeSystem.Types;
+using Bicep.Core.Workspaces;
+
+namespace Bicep.Core.Semantics.Namespaces;
+
+public record NamespaceResult(
+    string Name,
+    string ProviderName,
+    TypeSymbol Type,
+    ProviderDeclarationSyntax? Origin);
+
+public interface INamespaceProvider
+{
+    IEnumerable<NamespaceResult> GetNamespaces(
+        RootConfiguration rootConfig,
+        IFeatureProvider features,
+        IArtifactFileLookup artifactFileLookup,
+        BicepSourceFile sourceFile,
+        ResourceScope targetScope);
+}
```

Comment:
Just curious, was there a specific rationale for removing INamespaceProvider.cs? I think we've been following the convention to put interfaces in their own files.